cmp/compare_test.go

```diff
@@ -326,93 +355,117 @@ func comparerTests() []test {
 		y:     struct{ fmt.Stringer }{regexp.MustCompile("hello2")},
 		opts:  []cmp.Option{cmp.Comparer(func(x, y fmt.Stringer) bool { return x.String() == y.String() })},
 		wantDiff: `
-root:
-	-: s"hello"
-	+: s"hello2"`,
+  struct{ fmt.Stringer }(
+- 	s"hello",
++ 	s"hello2",
+  )
+`,
 	}, {
 		label: label,
 		x:     md5.Sum([]byte{'a'}),
 		y:     md5.Sum([]byte{'b'}),
 		wantDiff: `
-{[16]uint8}[0]:
-	-: 0x0c
-	+: 0x92
-{[16]uint8}[1]:
-	-: 0xc1
-	+: 0xeb
-{[16]uint8}[2]:
-	-: 0x75
-	+: 0x5f
-{[16]uint8}[3]:
-	-: 0xb9
-	+: 0xfe
-{[16]uint8}[4]:
-	-: 0xc0
-	+: 0xe6
-{[16]uint8}[5]:
-	-: 0xf1
-	+: 0xae
-{[16]uint8}[6]:
-	-: 0xb6
-	+: 0x2f
-{[16]uint8}[7]:
-	-: 0xa8
-	+: 0xec
-{[16]uint8}[8]:
-	-: 0x31
-	+: 0x3a
-{[16]uint8}[9]:
-	-: 0xc3
-	+: 0xd7
-{[16]uint8}[10]:
-	-: 0x99
-	+: 0x1c
-{[16]uint8}[11->?]:
-	-: 0xe2
-	+: <non-existent>
-{[16]uint8}[12->?]:
-	-: 0x69
-	+: <non-existent>
-{[16]uint8}[?->12]:
-	-: <non-existent>
-	+: 0x75
-{[16]uint8}[?->13]:
-	-: <non-existent>
-	+: 0x31
-{[16]uint8}[14]:
-	-: 0x26
-	+: 0x57
-{[16]uint8}[15]:
-	-: 0x61
-	+: 0x8f`,
+  [16]uint8{
+- 	0x0c,
++ 	0x92,
+- 	0xc1,
++ 	0xeb,
+- 	0x75,
++ 	0x5f,
+- 	0xb9,
++ 	0xfe,
+- 	0xc0,
++ 	0xe6,
+- 	0xf1,
++ 	0xae,
+- 	0xb6,
++ 	0x2f,
+- 	0xa8,
++ 	0xec,
+- 	0x31,
++ 	0x3a,
+- 	0xc3,
++ 	0xd7,
+- 	0x99,
++ 	0x1c,
+- 	0xe2,
+- 	0x69,
+  	0x77,
++ 	0x75,
++ 	0x31,
+- 	0x26,
++ 	0x57,
+- 	0x61,
++ 	0x8f,
+  }
+`,
 	}, {
 		label: label,
 		x:     new(fmt.Stringer),
 		y:     nil,
 		wantDiff: `
-root:
-	-: &<nil>
-	+: <non-existent>`,
+  interface{}(
+- 	&fmt.Stringer(nil),
+  )
+`,
 	}, {
 		label: label,
 		x:     makeTarHeaders('0'),
 		y:     makeTarHeaders('\x00'),
 		wantDiff: `
-{[]cmp_test.tarHeader}[0].Typeflag:
-	-: 0x30
-	+: 0x00
-{[]cmp_test.tarHeader}[1].Typeflag:
-	-: 0x30
-	+: 0x00
-{[]cmp_test.tarHeader}[2].Typeflag:
-	-: 0x30
-	+: 0x00
-{[]cmp_test.tarHeader}[3].Typeflag:
-	-: 0x30
-	+: 0x00
-{[]cmp_test.tarHeader}[4].Typeflag:
-	-: 0x30
-	+: 0x00`,
+  []cmp_test.tarHeader{
```

Comment:
I prefer that approach as well, but it turns out to make the text formatting more complicated since the current logic has a heuristic where it always prints equal nodes as a single line, while it expands and different nodes. Single-line collapsing doesn't work well with that style of formatting.

A future PR may add some intelligence about when it can collapse slices in that way.